controllers/storageclusterpeer/storageclusterpeer_controller.go

```diff
@@ -18,45 +18,285 @@ package storageclusterpeer
+	if err := r.list(r.cephBlockPoolList, client.InNamespace(r.storageClusterPeer.Namespace)); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	r.clusterID = util.GetClusterID(r.ctx, r.Client, &r.Log)
+	if r.clusterID == "" {
+		err := fmt.Errorf("failed to get clusterID from the ClusterVersion CR")
+		r.Log.Error(err, "failed to get ClusterVersion of the OCP Cluster")
+		return reconcile.Result{}, err
+	}
+	var err error
+
+	r.ocsClient, err = r.newExternalClient()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	defer r.ocsClient.Close()
+
+	var result reconcile.Result
+	var reconcileError error
+
+	result, reconcileError = r.reconcilePhases()
+
+	// Apply status changes to the StorageClusterPeer
+	statusError := r.Client.Status().Update(r.ctx, r.storageClusterPeer)
+	if statusError != nil {
+		r.Log.Info("Failed to update StorageClusterPeer status.")
+	}
 
-	// TODO(user): your logic here
+	// Reconcile errors have higher priority than status update errors
+	if reconcileError != nil {
+		return result, reconcileError
+	}
 
-	return ctrl.Result{}, nil
+	if statusError != nil {
+		return result, statusError
+	}
+
+	return result, nil
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *StorageClusterPeerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	enqueueStorageClusterPeerRequest := handler.EnqueueRequestsFromMapFunc(
+		func(ctx context.Context, obj client.Object) []reconcile.Request {
+			// Get the StorageClusterPeer objects
+			scpList := &ocsv1.StorageClusterPeerList{}
+			err := r.Client.List(ctx, scpList, &client.ListOptions{Namespace: obj.GetNamespace()})
+			if err != nil {
+				r.Log.Error(err, "Unable to list StorageClusterPeer objects")
+				return []reconcile.Request{}
+			}
+
+			// Return name and namespace of the StorageClusterPeers object
+			request := []reconcile.Request{}
+			for _, scp := range scpList.Items {
+				request = append(request, reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Namespace: scp.Namespace,
+						Name:      scp.Name,
+					},
+				})
+			}
+
+			return request
+		},
+	)
+
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&ocsv1.StorageClusterPeer{}).
+		For(&ocsv1.StorageClusterPeer{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(&rookCephv1.CephBlockPool{}, enqueueStorageClusterPeerRequest).
 		Complete(r)
 }
+
+func (r *StorageClusterPeerReconciler) reconcilePhases() (reconcile.Result, error) {
+	r.Log.Info("Running StorageClusterPeer controller")
+
+	//marked for deletion
+	if !r.storageClusterPeer.GetDeletionTimestamp().IsZero() {
+		r.storageClusterPeer.Status.Phase = ocsv1.StorageClusterPeerDeleting
+
+		if err := r.reconcileRevokeBlockPoolPeering(); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		//remove finalizer
+		r.Log.Info("removing finalizer from the StorageClusterPeer resource")
+		controllerutil.RemoveFinalizer(r.storageClusterPeer, ocsv1.StorageClusterPeerFinalizer)
```

Comment:
incorrect I guess, you are removing finalizer from in-memory object but not on api server, need to update here.